File: samples/da-trey-research-python/src/services/project_api_service.py
from datetime import datetime

from ..model.base_model import Project
from ..model.api_model import ApiProject, ApiProjectAssignment, ApiAddConsultantToProjectResponse
from .project_db_service import project_db_service
from .assignment_db_service import assignment_db_service
from .consultant_db_service import consultant_db_service
from .utilities import HttpError


class ProjectApiService:
    """API service for project operations."""

    def get_api_project_by_id(self, project_id: str) -> ApiProject:
        project = project_db_service.get_project_by_id(project_id)
        assignments = assignment_db_service.get_assignments()
        return self._get_api_project(project, assignments)

    def get_api_projects(self, project_or_client_name: str, consultant_name: str) -> list:
        projects = project_db_service.get_projects()
        assignments = assignment_db_service.get_assignments()

        # Filter on base properties
        if project_or_client_name:
            projects = [
                p for p in projects
                if project_or_client_name.lower() in p.name.lower()
                or project_or_client_name.lower() in p.clientName.lower()
            ]

        # Remove duplicates
        seen_ids = set()
        unique_projects = []
        for p in projects:
            if p.id not in seen_ids:
                seen_ids.add(p.id)
                unique_projects.append(p)
        projects = unique_projects

        # Augment with assignment information
        result = [self._get_api_project(p, assignments) for p in projects]

        # Filter on consultant name
        if result and consultant_name:
            result = [
                p for p in result
                if any(consultant_name.lower() in c.consultantName.lower() for c in p.consultants)
            ]

        return result

    def _get_api_project(self, project: Project, assignments: list) -> ApiProject:
        api_project = ApiProject(
            id=project.id,
            name=project.name,
            description=project.description,
            clientName=project.clientName,
            clientContact=project.clientContact,
            clientEmail=project.clientEmail,
            location=project.location,
            mapUrl=project.mapUrl,
            consultants=[],
            forecastThisMonth=0,
            forecastNextMonth=0,
            deliveredLastMonth=0,
            deliveredThisMonth=0,
        )

        project_assignments = [a for a in assignments if a.projectId == project.id]

        for assignment in project_assignments:
            consultant = consultant_db_service.get_consultant_by_id(assignment.consultantId)
            forecast_hours = self._find_hours(assignment.forecast)
            delivered_hours = self._find_hours(assignment.delivered)

            api_project.consultants.append(ApiProjectAssignment(
                consultantName=consultant.name,
                consultantLocation=consultant.location,
                role=assignment.role,
                forecastThisMonth=forecast_hours["thisMonth"],
                forecastNextMonth=forecast_hours["nextMonth"],
                deliveredLastMonth=delivered_hours["lastMonth"],
                deliveredThisMonth=delivered_hours["thisMonth"],
            ))
            api_project.forecastThisMonth += forecast_hours["thisMonth"]
            api_project.forecastNextMonth += forecast_hours["nextMonth"]
            api_project.deliveredLastMonth += delivered_hours["lastMonth"]
            api_project.deliveredThisMonth += delivered_hours["thisMonth"]

        return api_project

    def _find_hours(self, hours: list) -> dict:
        """Extract last, this, and next month's hours from an array of HoursEntry dicts."""
        now = datetime.now()
        this_month = now.month
        this_year = now.year

        last_month = 12 if this_month == 1 else this_month - 1
        last_year = this_year - 1 if this_month == 1 else this_year

        next_month = 1 if this_month == 12 else this_month + 1
        next_year = this_year + 1 if this_month == 12 else this_year

        last_month_hours = 0.0
        this_month_hours = 0.0
        next_month_hours = 0.0

        if hours:
            for h in hours:
                m = h.get("month", 0) if isinstance(h, dict) else h.month
                y = h.get("year", 0) if isinstance(h, dict) else h.year
                hrs = h.get("hours", 0) if isinstance(h, dict) else h.hours
                if m == last_month and y == last_year:
                    last_month_hours = hrs
                elif m == this_month and y == this_year:
                    this_month_hours = hrs
                elif m == next_month and y == next_year:
                    next_month_hours = hrs

        return {
            "lastMonth": last_month_hours,
            "thisMonth": this_month_hours,
            "nextMonth": next_month_hours,
        }

    def add_consultant_to_project(
        self, project_name: str, consultant_name: str, role: str, hours: float
    ) -> ApiAddConsultantToProjectResponse:
        from .consultant_api_service import consultant_api_service

        projects = self.get_api_projects(project_name, "")
        consultants = consultant_api_service.get_api_consultants(consultant_name, "", "", "", "", "")

        if not projects:
            raise HttpError(404, f"Project not found: {project_name}")
        elif len(projects) > 1:
            raise HttpError(406, f"Multiple projects found with the name: {project_name}")
        elif not consultants:
            raise HttpError(404, f"Consultant not found: {consultant_name}")
        elif len(consultants) > 1:
            raise HttpError(406, f"Multiple consultants found with the name: {consultant_name}")

        project = projects[0]
        consultant = consultants[0]

        remaining_forecast = assignment_db_service.add_consultant_to_project(
            project.id, consultant.id, role, hours
        )
        message = (
            f'Added consultant {consultant.name} to {project.clientName} on project '
            f'"{project.name}" with {remaining_forecast} hours forecast this month.'
        )

        return ApiAddConsultantToProjectResponse(
            clientName=project.clientName,
            projectName=project.name,
            consultantName=consultant.name,
            remainingForecast=remaining_forecast,
            message=message,
        )


# Singleton instance
project_api_service = ProjectApiService()

File: samples/da-trey-research-cs/Services/ProjectApiService.cs
using TreyResearch.Models;

namespace TreyResearch.Services;

public class ProjectApiService
{
    private static readonly ProjectApiService _instance = new();
    public static ProjectApiService Instance => _instance;

    public async Task<ApiProject> GetApiProjectByIdAsync(string projectId)
    {
        var project = await ProjectDbService.Instance.GetProjectByIdAsync(projectId);
        var assignments = await AssignmentDbService.Instance.GetAssignmentsAsync();
        return await GetApiProjectAsync(project, assignments);
    }

    public async Task<List<ApiProject>> GetApiProjectsAsync(string projectOrClientName, string consultantName)
    {
        var projects = await ProjectDbService.Instance.GetProjectsAsync();
        var assignments = await AssignmentDbService.Instance.GetAssignmentsAsync();

        // Filter on base properties
        if (!string.IsNullOrEmpty(projectOrClientName))
        {
            projects = projects.Where(p =>
                p.Name.Contains(projectOrClientName, StringComparison.OrdinalIgnoreCase) ||
                p.ClientName.Contains(projectOrClientName, StringComparison.OrdinalIgnoreCase)
            ).ToList();
        }

        // Remove duplicates
        projects = projects.DistinctBy(p => p.Id).ToList();

        // Augment with assignment information
        var result = new List<ApiProject>();
        foreach (var p in projects)
        {
            result.Add(await GetApiProjectAsync(p, assignments));
        }

        // Filter on consultant name
        if (!string.IsNullOrEmpty(consultantName))
        {
            result = result.Where(p =>
                p.Consultants.Any(c => c.ConsultantName.Contains(consultantName, StringComparison.OrdinalIgnoreCase))
            ).ToList();
        }

        return result;
    }

    private async Task<ApiProject> GetApiProjectAsync(Project project, List<Assignment> assignments)
    {
        var result = new ApiProject
        {
            Id = project.Id,
            Name = project.Name,
            Description = project.Description,
            ClientName = project.ClientName,
            ClientContact = project.ClientContact,
            ClientEmail = project.ClientEmail,
            Location = project.Location,
            MapUrl = project.MapUrl
        };

        var projectAssignments = assignments.Where(a => a.ProjectId == project.Id).ToList();

        foreach (var assignment in projectAssignments)
        {
            var consultant = await ConsultantDbService.Instance.GetConsultantByIdAsync(assignment.ConsultantId);
            var (forecastLastMonth, forecastThisMonth, forecastNextMonth) = FindHours(assignment.Forecast);
            var (deliveredLastMonth, deliveredThisMonth, deliveredNextMonth) = FindHours(assignment.Delivered);

            result.Consultants.Add(new ApiProjectAssignment
            {
                ConsultantName = consultant.Name,
                ConsultantLocation = consultant.Location,
                Role = assignment.Role,
                ForecastThisMonth = forecastThisMonth,
                ForecastNextMonth = forecastNextMonth,
                DeliveredLastMonth = deliveredLastMonth,
                DeliveredThisMonth = deliveredThisMonth
            });

            result.ForecastThisMonth += forecastThisMonth;
            result.ForecastNextMonth += forecastNextMonth;
            result.DeliveredLastMonth += deliveredLastMonth;
            result.DeliveredThisMonth += deliveredThisMonth;
        }

        return result;
    }

    private static (double lastMonth, double thisMonth, double nextMonth) FindHours(List<HoursEntry> hours)
    {
        var now = DateTime.Now;
        var thisMonth = now.Month;
        var thisYear = now.Year;

        var lastMonth = thisMonth == 1 ? 12 : thisMonth - 1;
        var lastYear = thisMonth == 1 ? thisYear - 1 : thisYear;

        var nextMonth = thisMonth == 12 ? 1 : thisMonth + 1;
        var nextYear = thisMonth == 12 ? thisYear + 1 : thisYear;

        var lastMonthHours = hours.FirstOrDefault(h => h.Month == lastMonth && h.Year == lastYear)?.Hours ?? 0;
        var thisMonthHours = hours.FirstOrDefault(h => h.Month == thisMonth && h.Year == thisYear)?.Hours ?? 0;
        var nextMonthHours = hours.FirstOrDefault(h => h.Month == nextMonth && h.Year == nextYear)?.Hours ?? 0;

        return (lastMonthHours, thisMonthHours, nextMonthHours);
    }

    public async Task<ApiAddConsultantToProjectResponse> AddConsultantToProjectAsync(
        string projectName, string consultantName, string role, double hours)
    {
        var projects = await GetApiProjectsAsync(projectName, "");
        var consultants = await ConsultantApiService.Instance.GetApiConsultantsAsync(consultantName, "", "", "", "", "");

        if (projects.Count == 0)
            throw new HttpError(404, $"Project not found: {projectName}");
        if (projects.Count > 1)
            throw new HttpError(406, $"Multiple projects found with the name: {projectName}");
        if (consultants.Count == 0)
            throw new HttpError(404, $"Consultant not found: {consultantName}");
        if (consultants.Count > 1)
            throw new HttpError(406, $"Multiple consultants found with the name: {consultantName}");

        var project = projects[0];
        var consultant = consultants[0];

        var remainingForecast = await AssignmentDbService.Instance.AddConsultantToProjectAsync(
            project.Id, consultant.Id, role, hours);

        var message = $"Added consultant {consultant.Name} to {project.ClientName} on project \"{project.Name}\" with {remainingForecast} hours forecast this month.";

        return new ApiAddConsultantToProjectResponse
        {
            ClientName = project.ClientName,
            ProjectName = project.Name,
            ConsultantName = consultant.Name,
            RemainingForecast = remainingForecast,
            Message = message
        };
    }
}
